openwrt-openwisp-monitoring/files/monitoring.agent

```diff
@@ -78,24 +65,140 @@ collect_data(){
+set_url_and_curl(){
+  if [ -z "$BASE_URL" ]; then
+    logger -s "missing required --url option" \
            -t openwisp_monitoring \
            -p daemon.err
+    exit 1
+  fi
 
-    [ "$verbose_mode" = "0" ] && logger -s "Run with verbose mode to find more." \
-                                        -t openwisp_monitoring \
-                                        -p daemon.err
+  if [ -z "$UUID" ]; then
+    logger -s "missing required --uuid option" \
+            -t openwisp_monitoring \
+            -p daemon.err
+    exit 1
   fi
-  sleep "$INTERVAL" & wait $!
-done
+
+  if [ -z "$KEY" ]; then
+    logger -s "missing required --key option" \
+           -t openwisp_monitoring \
+           -p daemon.err
+    exit 1
+  fi
+
+  URL="$BASE_URL/api/v1/monitoring/device/$UUID/?key=$KEY"
+
+  CURL_COMMAND="curl -s -w "%{http_code}""
+  [ "$VERIFY_SSL" = "0" ] && CURL_COMMAND="$CURL_COMMAND -k"
+  [ "$verbose_mode" = "1" ] && CURL_COMMAND="$CURL_COMMAND -v"
+  echo "0"
+}
+
+save_data() {
+  # Remove double quotes from interfaces
+  MONITORED_INTERFACES=$(echo "$MONITORED_INTERFACES" | tr -d '"')
+
+  while true
+  do
+    data="$(collect_data)"
+    memory_available="$(check_available_memory)"
+    if [ "$memory_available" != "0" ]; then
+        file_name="$(date -u +'%d-%m-%Y_%H:%M:%S').000000.json" # extra zeroes are added for nanoseconds precision
+        # make directory
+        mkdir -p "$TMP_DIR"
+        #save data with file_name
+        echo "$data" > "$TMP_DIR/$file_name"
+        [ "$verbose_mode" = "1" ] && logger -s "Data saved temporarily" \
+                                            -p daemon.info
+    fi
+    # get process id of container sending data
+    pid=$(pgrep -f "openwisp_monitoring.*uuid")
+    kill -SIGUSR1 "$pid"
+    sleep "$INTERVAL"
+  done
+}
+
+handle_sigusr1() {
+  if [ "$verbose_mode" = "0" ]; then
+    logger -s "SIGUSR1 received! Sending data" \
+           -t openwisp_monitoring \
+           -p daemon.info
+  else
+    logger -s "SIGUSR1 received! Sending data" \
+           -p daemon.info
+  fi
+}
+
+send_data() {
+  while true
+  do
+    for file in "$TMP_DIR"/*.json
+    do
+      if [ ! -f "$file" ]; then
+        [ "$verbose_mode" = "1" ] && logger -s "No data file found to send." \
+                                            -p daemon.info
+        trap handle_sigusr1 USR1
+        # SIGUSR1 signal received interrupt sleep and continue sending data
+        sleep "$INTERVAL" & wait $!
+        trap "" USR1
+        continue
+      fi
+      basefilename=${file##*/}
+      filename=${basefilename%.*}
+      url="$URL&time=$filename"
+      data=$(cat "$file")
+      # retry sending data in case of failure
+      MAX_RETRY=${MAX_RETRY:-5}
+      FAILURES=0
+      BASE_SECONDS=1
+      # check if the data is latest or old one
+      [ "$(echo "$TMP_DIR"/*.json | awk '{print $2}')" ] || url="$url&current=true"
+      while true
+      do
+        FAILURES=$((FAILURES+1))
```

Comment:
why are you adding 1 here, before any failure?
Is this `FAILURES` variable actually a counter of how many attempts have been made? In that case I would name it `attempts`.

Reply:
Changing in to increase after failure